src/mix_protocol.nim

```diff
@@ -49,123 +52,117 @@ proc cryptoRandomInt(max: int): Result[int, string] =
   let value = cast[uint64](bytes)
   return ok(int(value mod uint64(max)))
 
-proc handleMixNodeConnection(mixProto: MixProtocol, conn: Connection) {.async.} =
-  while true:
-    var receivedBytes = await conn.readLp(packetSize)
+proc handleMixNodeConnection(
+    mixProto: MixProtocol, conn: Connection
+) {.async: (raises: [CancelledError]).} =
+  var receivedBytes: seq[byte]
+  try:
+    receivedBytes = await conn.readLp(packetSize)
+  except Exception as e:
+    error "Failed to read: ", err = e.msg
+  finally:
+    if conn != nil:
+      try:
+        await conn.close()
+      except CatchableError as e:
+        error "Failed to close incoming stream: ", err = e.msg
 
-    if receivedBytes.len == 0:
-      break # No data, end of stream
+  if receivedBytes.len == 0:
+    return # No data, end of stream
 
-    # Process the packet
-    let (multiAddr, _, mixPrivKey, _, _) = getMixNodeInfo(mixProto.mixNodeInfo)
+  # Process the packet
+  let (multiAddr, _, mixPrivKey, _, _) = getMixNodeInfo(mixProto.mixNodeInfo)
 
-    let processedPktRes =
-      processSphinxPacket(receivedBytes, mixPrivKey, mixProto.tagManager)
-    if processedPktRes.isErr:
-      error "Failed to process Sphinx packet", err = processedPktRes.error
-      return
-    let (nextHop, delay, processedPkt, status) = processedPktRes.get()
+  let processedPktRes =
+    processSphinxPacket(receivedBytes, mixPrivKey, mixProto.tagManager)
+  if processedPktRes.isErr:
+    error "Failed to process Sphinx packet", err = processedPktRes.error
+    return
+  let (nextHop, delay, processedPkt, status) = processedPktRes.get()
 
-    case status
-    of Exit:
-      # This is the exit node, forward to destination
-      let msgChunk = deserializeMessageChunk(processedPkt).valueOr:
-        error "Deserialization failed", err = error
-        return
+  case status
+  of Exit:
+    if (nextHop != Hop()) or (delay != @[]):
+      error "Next hop and delay must be empty"
+      return
 
-      let unpaddedMsg = unpadMessage(msgChunk).valueOr:
-        error "Unpadding message failed", err = error
-        return
+    # This is the exit node, forward to destination
+    let msgChunk = deserializeMessageChunk(processedPkt).valueOr:
+      error "Deserialization failed", err = error
+      return
 
-      let deserializedResult = deserializeMixMessage(unpaddedMsg).valueOr:
-        error "Deserialization failed", err = error
-        return
+    let unpaddedMsg = unpadMessage(msgChunk).valueOr:
+      error "Unpadding message failed", err = error
+      return
 
-      let (message, protocol) = getMixMessage(deserializedResult)
-      info "# Received: ", receiver = multiAddr, message = message
+    let deserializedResult = deserializeMixMessage(unpaddedMsg).valueOr:
+      error "Deserialization failed", err = error
+      return
 
-      # Add delay
-      let delayMillis = (delay[0].int shl 8) or delay[1].int
-      await sleepAsync(milliseconds(delayMillis))
```

Comment:
do we need a deley when forwarding on to destination? My naive guess is the time-mixing is the purview of the intermediate nodes. if you want that extra delay at the exit, why not just have an extra intermediate? or is it that having it here gives you an extra layer of time-mixing "for free"?